### PR TITLE
Created a Kokkos version of the ML-IAP routines.  Completed one model…

### PR DESCRIPTION
**Pull request by bathmatt**
… and

one descriptor along with the pair style and data member.  To test the stack one can run

lmp -k on g 1 -sf kk -pk kokkos neigh full newton on -in in.mliap.so3.Ni_Mo -echo both

and compare it to the non-kokkos results.
Results are identical for cuda/openmp/serial and non-kokkos

**Summary**

This enabled kokkos for the ML-IAP routines, Used as a test of principal for how this can be done.  We plan to extend this to more useful python based models 

**Related Issue(s)**

**Author(s)**

Matt Bettencourt and Markus Hrywniak at NVIDIA

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by stanmoore1** [comment]
@bathmatt looks like some unit tests are failing with a segmentation fault:

```
99% tests passed, 5 tests failed out of 544

The following tests FAILED:
	313 - ManybodyPairStyle:mliap_nn (SEGFAULT)
	314 - ManybodyPairStyle:mliap_snap_chem (SEGFAULT)
	315 - ManybodyPairStyle:mliap_snap_linear (SEGFAULT)
	316 - ManybodyPairStyle:mliap_snap_quadratic (SEGFAULT)
	317 - ManybodyPairStyle:mliap_so3 (SEGFAULT)
Errors while running CTest
```

**Comment by bathmatt** [comment]
OK, is this with the kokkos or non-kokkos versions?  I can take a look

**Comment by stanmoore1** [comment]
It's Kokkos with OpenMP backend

**Comment by akohlmey** [comment]
@bathmatt please note that the segfaults could also happen due to missing features in your implementation. The pair_style test program cannot (yet) handle the case where some command throws an error during one of the script fragments provided by the test description files (in YAML format) in the unittest/force-styles/tests folder.

You should be able to see this when running the tests interactively with:
```
TEST_ARGS=-v ctest -R mliap -V
```

The test program is dependent on suitable introspection functionality to detect whether its commands would succeed or not, but that is incomplete and not always easy to achieve. That is the reason why we can test the KOKKOS package pair styles with OpenMP, but not yet with any kind of GPU (since it only depends on how LAMMPS is compiled and not on compatible hardware being present and a compatible driver).

**Comment by bathmatt** [comment]
Finally tracked down the issue.  There is a mismatch in the base code and the modified code between nelements and atom->ntypes.  Where the model thinks there is only 1 type of atom, atom->ntypes is 8.  Therefore there is a miscommunication on how big things like cutsq and other stuff should be.  

General question, why does everyone have their own copy of cutsq?  Is it pair type dependent?  Or just material property dependent?  Sorry, I don't know MD at all really

**Comment by akohlmey** [comment]
> General question, why does everyone have their own copy of cutsq? Is it pair type dependent? Or just material property dependent? Sorry, I don't know MD at all really

Yes, this is very model and thus pair style dependent. Generally keeping `cutsq` is an optimization, so you don't have to compute `cut*cut` or rather don't have to use `sqrt(rsq)` when checking for cutoffs.

For example with `lj/cut` you can have a different cutoff for each pair of atom types (and a global default), with `lj/cut/coul/cut`, you can have a different cutoff for every pair of atom types for *either* Lennard-Jones *or* Coulomb. With `lj/cut/coul/long`, there is *only* a *global* Coulomb cutoff allowed and with `lj/long/coul/long` only a single global cutoff for both LJ and Coul.

**Comment by bathmatt** [comment]
Resolved these issues.  Thanks for @akohlmey for explaining the mapping between pairs and descriptors for cutsq

```
(base) mattb@0cb28a7-lcedt:~/lammps/LAMMPS/omp-test$ ctest -R mliap -j 32
Test project /home/mattb/lammps/LAMMPS/omp-test
    Start 306: ManybodyPairStyle:mliap_snap_chem
    Start 309: ManybodyPairStyle:mliap_so3
    Start 308: ManybodyPairStyle:mliap_snap_quadratic
    Start 305: ManybodyPairStyle:mliap_nn
    Start 307: ManybodyPairStyle:mliap_snap_linear
1/5 Test #307: ManybodyPairStyle:mliap_snap_linear ......   Passed    0.96 sec
2/5 Test #305: ManybodyPairStyle:mliap_nn ...............   Passed    1.03 sec
3/5 Test #308: ManybodyPairStyle:mliap_snap_quadratic ...   Passed    1.14 sec
4/5 Test #309: ManybodyPairStyle:mliap_so3 ..............   Passed    2.55 sec
5/5 Test #306: ManybodyPairStyle:mliap_snap_chem ........   Passed    2.73 sec

100% tests passed, 0 tests failed out of 5

Label Time Summary:
slow        =   8.41 sec*proc (5 tests)
unstable    =   2.55 sec*proc (1 test)

Total Test time (real) =   2.76 sec
(base) mattb@0cb28a7-lcedt:~/lammps/LAMMPS/omp-test$ 
```

**Comment by stanmoore1** [comment]
@bathmatt great, once you push the changes here I will review the PR.

**Comment by bathmatt** [comment]
@stanmoore1 @akohlmey the tests should pass.  I will check later today.  But think we are ready to review now..

**Review by stanmoore1 — APPROVED** [review]

File: src/KOKKOS/mliap_descriptor_so3_kokkos.cpp [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+#define MAXLINE 1024
+#define MAXWORD 3
+
+/* ---------------------------------------------------------------------- */
+template <class DeviceType>
+MLIAPDescriptorSO3Kokkos<DeviceType>::MLIAPDescriptorSO3Kokkos(LAMMPS *lmp, char *paramfilename)
+ // TODO: why take self as param, shouldn't be needed
+  :  Pointers(lmp), MLIAPDescriptorSO3(lmp, paramfilename), MLIAPDescriptorKokkos<DeviceType>(lmp, this)
+{
+  // TODO: the MLIAP_SO3 object likely needs a kokkos-ified version
```
Comment (stanmoore1) [inline src/KOKKOS/mliap_descriptor_so3_kokkos.cpp:43]: @bathmatt is this comment outdated?

File: src/KOKKOS/pair_mliap_kokkos.cpp [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+  if (vflag_atom && k_vatom.h_view.extent(0) < maxeatom) {
+    memoryKK->destroy_kokkos(k_vatom,vatom);
+    memoryKK->create_kokkos(k_vatom,vatom,maxeatom,6,"pair:eatom");
+  }
+
+  data->generate_neighdata(list, eflag, vflag);
+
+  // compute descriptors, if needed
+  if (!is_kokkos_descriptor) {
+    // put sync stuff here
```
Comment (stanmoore1) [inline src/KOKKOS/pair_mliap_kokkos.cpp]: Missing code from comment?

**Comment by stanmoore1** [comment]
@bathmatt For CUDA, this appears to only work with UVM enabled. Probably a missing `sync` or `modify`.

**Comment by akohlmey** [comment]
@stanmoore1 are you done with updates to this PR? I would like to merge it with upstream and apply the homepage/email fixes so that the "checkstyle" test would no longer fail.

**Comment by stanmoore1** [comment]
@akohlmey I'm done for now. I'm going to do another review with a few change requests, but I can wait if necessary.

File: src/KOKKOS/pair_mliap_kokkos.cpp [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
```
Comment (stanmoore1) [inline src/KOKKOS/pair_mliap_kokkos.cpp]: Please add sync or clean up code if not needed.

File: src/KOKKOS/mliap_model_linear_kokkos.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  k_data->modified(execution_space, BETAS_MASK | EATOMS_MASK);
+}
+
+/* ---------------------------------------------------------------------- */
+
+template<class DeviceType>
+void MLIAPModelLinearKokkos<DeviceType>::compute_gradgrads(class MLIAPData *data)
+{
+  abort();
```
Comment (stanmoore1) [inline src/KOKKOS/mliap_model_linear_kokkos.cpp]: Please use the LAMMPS `Error` class and put in a suitable error message.

File: src/KOKKOS/mliap_so3_kokkos.cpp [hunk 5]
```diff
@@ -0,0 +1,1196 @@
...
+template <class DeviceType>
+void MLIAP_SO3Kokkos<DeviceType>::spectrum(int nlocal, DAT::tdual_int_1d numneighs, DAT::tdual_int_1d jelems, DAT::tdual_float_1d wjelem,
+                               DAT::tdual_float_2d rij,  DAT::tdual_int_1d k_ij,
+                               int nmax, int lmax, double rcut, double alpha, int totaln, int ncoefs)
+{
+  init_arrays(nlocal, ncoefs);
+
+  bigint totali;
+
+  totali = totaln * m_Nmax * (m_lmax + 1);
```
Comment (stanmoore1) [inline src/KOKKOS/mliap_so3_kokkos.cpp:735]: Need to add a check to only realloc if the size has been exceeded. Please also disable initialization if zeroing is not needed.

File: src/KOKKOS/mliap_descriptor_so3_kokkos.cpp [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+      Kokkos::atomic_add(&d_f(i, 0),fij[0]);
+      Kokkos::atomic_add(&d_f(i, 1),fij[1]);
+      Kokkos::atomic_add(&d_f(i, 2),fij[2]);
+      Kokkos::atomic_add(&d_f(j, 0),-fij[0]);
+      Kokkos::atomic_add(&d_f(j, 1),-fij[1]);
+      Kokkos::atomic_add(&d_f(j, 2),-fij[2]);
+
+      // add in global and per-atom virial contributions
+      // this is optional and has no effect on force calculation
+      // FIXME: v_tally not implemented on pair mliap kokkos
```
Comment (stanmoore1) [inline src/KOKKOS/mliap_descriptor_so3_kokkos.cpp]: Remove `FIXME` or convert to `TO-DO`

File: src/KOKKOS/mliap_so3_kokkos.cpp [hunk 7]
```diff
@@ -0,0 +1,1196 @@
...
+    }
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+template <class DeviceType>
+KOKKOS_INLINE_FUNCTION
+void MLIAP_SO3Kokkos<DeviceType>::operator() (const MLIAP_SO3Kokkos<DeviceType>::MLIAPSO3SpectrumDXDRTag&, int ii) const {
+  //FIXME Need to move m_ulist_r, m_ulist_i into local shared memory
```
Comment (stanmoore1) [inline src/KOKKOS/mliap_so3_kokkos.cpp]: Finish `FIXME` or convert to `TO-DO`

File: src/KOKKOS/mliap_descriptor_so3_kokkos.cpp [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+#define MAXLINE 1024
+#define MAXWORD 3
+
+/* ---------------------------------------------------------------------- */
+template <class DeviceType>
+MLIAPDescriptorSO3Kokkos<DeviceType>::MLIAPDescriptorSO3Kokkos(LAMMPS *lmp, char *paramfilename)
+ // TODO: why take self as param, shouldn't be needed
+  :  Pointers(lmp), MLIAPDescriptorSO3(lmp, paramfilename), MLIAPDescriptorKokkos<DeviceType>(lmp, this)
+{
+  // TODO: the MLIAP_SO3 object likely needs a kokkos-ified version
```
Comment (stanmoore1) [inline src/KOKKOS/mliap_descriptor_so3_kokkos.cpp:43]: Double check this comment, it looks outdated.

**Review by stanmoore1 — CHANGES_REQUESTED** [review]
@bathmatt I put in a few change requests. Also please double check CUDA, it appears to be giving the wrong numerics. I checked and it was failing in the version before I made changes. It works with UVM but not without.

**Comment by akohlmey** [comment]
> @akohlmey I'm done for now. I'm going to do another review with a few change requests, but I can wait if necessary.

back to you.

**Comment by bathmatt** [comment]
> @bathmatt For CUDA, this appears to only work with UVM enabled. Probably a missing `sync` or `modify`.

So, i tested with a kokkos model only and a kokkos model and a kokkos descriptor and those work.  I didn't test with kokkos descritpor and CPU model.  let me insure that works and address your other comment.  They might be related.  

what did you do to show that UVM was required?  Looking for a test

**Comment by bathmatt** [comment]
@stanmoore1 I built the unit tests with cuda and ran the tests and without UVM it passed.  I am guessing your comment might be related to the comment in the code.  I will verify that the comment has been addressed or address those concerns  but 

```
(base) mattb@0cb28a7-lcedt:~/lammps/LAMMPS/cuda-tests$ ctest -R mliap
Test project /home/mattb/lammps/LAMMPS/cuda-tests
    Start 305: ManybodyPairStyle:mliap_nn
1/5 Test #305: ManybodyPairStyle:mliap_nn ...............   Passed    0.95 sec
    Start 306: ManybodyPairStyle:mliap_snap_chem
2/5 Test #306: ManybodyPairStyle:mliap_snap_chem ........   Passed    1.87 sec
    Start 307: ManybodyPairStyle:mliap_snap_linear
3/5 Test #307: ManybodyPairStyle:mliap_snap_linear ......   Passed    0.78 sec
    Start 308: ManybodyPairStyle:mliap_snap_quadratic
4/5 Test #308: ManybodyPairStyle:mliap_snap_quadratic ...   Passed    0.89 sec
    Start 309: ManybodyPairStyle:mliap_so3
5/5 Test #309: ManybodyPairStyle:mliap_so3 ..............   Passed    1.83 sec

100% tests passed, 0 tests failed out of 5

Label Time Summary:
slow        =   6.32 sec*proc (5 tests)
unstable    =   1.83 sec*proc (1 test)

Total Test time (real) =   6.34 sec
```
This is for the cmake here
```
(base) mattb@0cb28a7-lcedt:~/lammps/LAMMPS/cuda-tests$ cmake -C ../cmake/presets/basic.cmake -C ../cmake/presets/kokkos-cuda.cmake -D BUILD_SHARED_LIBS=on -D LAMMPS_EXCEPTIONS=on -D PKG_PYTHON=on -DPKG_ML-SNAP=ON  -DPKG_KOKKOS=ON -DPKG_ML-IAP=ON -DKokkos_ENABLE_CUDA_UVM=OFF -DCMAKE_BUILD_TYPE=DEBUG -DKokkos_ARCH_AMPERE86=ON -DKokkos_ARCH_PASCAL60=OFF  ../cmake/ -D ENABLE_TESTING=on

```

**Comment by akohlmey** [comment]
> @stanmoore1 I built the unit tests with cuda and ran the tests and without UVM it passed.

Units tests will *not* run KOKKOS styles with GPU acceleration, only KOKKOS with OpenMP is supported. This is due to a lack of introspection that can assure the existence of both, a properly compiled LAMMPS executable and a compatible GPU. I think this was mentioned before. So the unit tests are useless in this regard.

Instead you should try running the examples in the examples/mliap folder (and compare the output).

**Comment by bathmatt** [comment]
> 
> Instead you should try running the examples in the examples/mliap folder (and compare the output).

Is there an automated way to do this?  I've checked a couple of the inputs but will try to test the rest that run with openmp.

**Comment by akohlmey** [comment]
> Is there an automated way to do this? I've checked a couple of the inputs but will try to test the rest that run with openmp.

Not through ctest. But it is trivial to write a shell script that loops over the inputs with different settings and checks the log files for completion of each run.

**Comment by bathmatt** [comment]
Is there an easy way to revert the kokkos update so I can run cuda+debug w/o that annoying kokkos bug?

**Comment by bathmatt** [comment]
I have a branch running without the kokkos mods and my changes so I'm progressing.  Nice catch @stanmoore1 I hope to have a PR by the close of today.

File: src/KOKKOS/mliap_descriptor_so3_kokkos.cpp (hunk 6 above)
Comment (bathmatt) [reply src/KOKKOS/mliap_descriptor_so3_kokkos.cpp]: fixed

File: src/KOKKOS/mliap_model_linear_kokkos.cpp (hunk 4 above)
Comment (bathmatt) [reply src/KOKKOS/mliap_model_linear_kokkos.cpp]: not needed anymore

File: src/KOKKOS/mliap_so3_kokkos.cpp (hunk 5 above)
Comment (bathmatt) [reply src/KOKKOS/mliap_so3_kokkos.cpp:735]: Sorry, copied the same code that is in the base, fixed this some places, resolved

File: src/KOKKOS/mliap_descriptor_so3_kokkos.cpp (hunk 8 above)
Comment (bathmatt) [reply src/KOKKOS/mliap_descriptor_so3_kokkos.cpp:43]: was

**Comment by stanmoore1** [comment]
I confirm correct numerics on GPUs.

**Comment by stanmoore1** [comment]
@athomps this touches a few core (non-Kokkos) ML-IAP package files, can you review as well?

**Review by stanmoore1 — APPROVED** [review]
I approve

**Review by athomps — APPROVED** [review]
I approve.

**Pull request merged**